[JIT][read path] Make NoneType annotation_str emit `NoneType` instead of `None`

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54746 [JIT][write path] Make NoneType annotation_str emit `NoneType` instead of `None`
* #54745 [JIT][write path] Fix TupleType.annotation_str to conform to `typing` module syntax for empty tuple type
* **#54642 [JIT][read path] Make NoneType annotation_str emit `NoneType` instead of `None`**
* #54641 [JIT][read path] Fix TupleType.annotation_str to conform to `typing` module syntax for empty tuple type

Previously, `NoneType::annotation_str` would return `"None"`. This is technically not correct as it refers to `None` and not `None`'s type. Example from the typing module:

```
>>> import typing
>>> l = typing.List[None]
>>> print(l)
typing.List[NoneType]
>>> l.__args__[0] is None
False
```

This PR updates `NoneType::annotation_str` to return `NoneType`. It also updates the lexer and schema type parser to accept `NoneType` as an annotation. This should be backward compatible since the old `None` syntax is still accepted on the read path

Differential Revision: [D27314174](https://our.internmc.facebook.com/intern/diff/D27314174)

Commit:
[JIT] Make NoneType annotation_str emit `NoneType` instead of `None`

[ghstack-poisoned]

Commit:
Update on "[JIT] Make NoneType annotation_str emit `NoneType` instead of `None`"


Previously, `NoneType::annotation_str` would return `"None"`. This is technically not correct as it refers to `None` and not `None`'s type. Example from the typing module:

```
>>> import typing
>>> l = typing.List[None]
>>> print(l)
typing.List[NoneType]
>>> l.__args__[0] is None
False
```

This PR updates `NoneType::annotation_str` to return `NoneType`. It also updates the lexer and schema type parser to accept `NoneType` as an annotation. This should be backward compatible since the old `None` syntax is still accepted on the read path

Differential Revision: [D27314174](https://our.internmc.facebook.com/intern/diff/D27314174)

[ghstack-poisoned]

Commit:
Update on "[JIT] Make NoneType annotation_str emit `NoneType` instead of `None`"


Previously, `NoneType::annotation_str` would return `"None"`. This is technically not correct as it refers to `None` and not `None`'s type. Example from the typing module:

```
>>> import typing
>>> l = typing.List[None]
>>> print(l)
typing.List[NoneType]
>>> l.__args__[0] is None
False
```

This PR updates `NoneType::annotation_str` to return `NoneType`. It also updates the lexer and schema type parser to accept `NoneType` as an annotation. This should be backward compatible since the old `None` syntax is still accepted on the read path

Differential Revision: [D27314174](https://our.internmc.facebook.com/intern/diff/D27314174)

[ghstack-poisoned]

SplitInfinity reviewed (commented):
Nice! My only concern is forward compatibility (old internal PyTorch binaries trying to load models produced using a version of PyTorch with this change included. The usual solution is to land the read path changes first, wait a while, and then land the write path changes.

Commit:
Update on "[JIT] Make NoneType annotation_str emit `NoneType` instead of `None`"


Previously, `NoneType::annotation_str` would return `"None"`. This is technically not correct as it refers to `None` and not `None`'s type. Example from the typing module:

```
>>> import typing
>>> l = typing.List[None]
>>> print(l)
typing.List[NoneType]
>>> l.__args__[0] is None
False
```

This PR updates `NoneType::annotation_str` to return `NoneType`. It also updates the lexer and schema type parser to accept `NoneType` as an annotation. This should be backward compatible since the old `None` syntax is still accepted on the read path

Differential Revision: [D27314174](https://our.internmc.facebook.com/intern/diff/D27314174)

[ghstack-poisoned]

Commit:
Update on "[JIT][read path] Make NoneType annotation_str emit `NoneType` instead of `None`"


Previously, `NoneType::annotation_str` would return `"None"`. This is technically not correct as it refers to `None` and not `None`'s type. Example from the typing module:

```
>>> import typing
>>> l = typing.List[None]
>>> print(l)
typing.List[NoneType]
>>> l.__args__[0] is None
False
```

This PR updates `NoneType::annotation_str` to return `NoneType`. It also updates the lexer and schema type parser to accept `NoneType` as an annotation. This should be backward compatible since the old `None` syntax is still accepted on the read path

Differential Revision: [D27314174](https://our.internmc.facebook.com/intern/diff/D27314174)

[ghstack-poisoned]
